### PR TITLE
feat(claim): claim a gift on egghead next

### PR DESCRIPTION
**Pull request by joelhooks**
https://www.loom.com/share/73f653609e1f45799d3cc635584dca5c

Ability to claim a gift.

- [ ] improve UX 
- [ ] merge https://github.com/eggheadio/egghead-rails/pull/4955

gifts

### CODE REVIEW

File: next.config.js [hunk 1]
```diff
@@ -227,11 +227,6 @@ const legacyRoutes = [
     destination: `${appUrl}/admin/:all*`,
     permanent: true,
   },
-  {
-    source: `/gifts/:all*`,
-    destination: `${appUrl}/gifts/:all*`,
-    permanent: true,
-  },
```
Comment (joelhooks) [inline next.config.js:234]: this redirect sends folks to app.egghead.io to claim their gift

File: src/pages/gifts/claim/[guid].tsx [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    if (!gift) {
+      return
+    }
+    const {claim_url} = gift
+
+    await axios.post(claim_url)
+
+    refreshUser()
+
+    router.push('/')
```
Comment (joelhooks) [inline src/pages/gifts/claim/[guid].tsx:60]: the viewer/user doesn't get refreshed here and there is no messaging that it succeeded

File: src/components/pages/home/index.tsx [hunk 3]
```diff
@@ -11,15 +12,35 @@ import Link from 'next/link'
...
+
+  React.useEffect(() => {
+    const {query} = router
+    if (query.message) {
+      toast(query.message as string, {
+        icon: '✅',
+        duration: 5000,
+      })
+    }
+  }, [router])
```
Comment (zacjones93) [inline src/components/pages/home/index.tsx:28]: `/learn` page checks if a message is set in the url and displays a toast pop up if so.

Currently set toast to 5 seconds, maybe needs to be lower?

File: src/components/pages/home/index.tsx [hunk 4]
```diff
@@ -11,15 +12,35 @@ import Link from 'next/link'
...
+
+  React.useEffect(() => {
+    const {query} = router
+    if (query.type === 'claimed') {
+      router.reload()
+    }
+  }, [])
```
Comment (zacjones93) [inline src/components/pages/home/index.tsx:35]: When a gift is claimed, we set a `type` in the url to `claimed` which will force a reload for the user so they can see the content they now have access to

File: src/pages/gifts/claim/[guid].tsx [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            <h1>{error.message}</h1>
+            {error.type === 'login' ? (
+              <p>
+                <Link href="/login">
+                  <a>Click here</a>
+                </Link>{' '}
+                to sign in. Once you are signed in you will need to click your
+                gift claim link again.
+              </p>
+            ) : null}
```
Comment (zacjones93) [inline src/pages/gifts/claim/[guid].tsx:79]: We only want to show the login message if the user is not logged in.

File: src/pages/gifts/claim/[guid].tsx [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+              </p>
+            )}
+            <PrimaryButton
+              url={{
+                pathname: '/learn/',
+                query: {
+                  message: 'You have claimed the membership coupon.',
+                  type: 'claimed',
+                },
+              }}
```
Comment (zacjones93) [inline src/pages/gifts/claim/[guid].tsx:121]: `Link` can accept an object with `pathname` and `query` on it so it will convert any strings to be url safe

**Review by zacjones93 — APPROVED** [review]

**Pull request merged**